onnxruntime/core/mlas/lib/qgemm.cpp

```diff
@@ -2760,67 +2809,42 @@ Return Value:
 
 --*/
 {
-    const size_t M = Parameters->M;
-    const size_t N = Parameters->N;
-    const size_t K = Parameters->K;
-
-    //
-    // Compute the number of target threads given the complexity of the SGEMM
-    // operation. Small requests should run using the single threaded path.
-    //
-
-    const double Complexity = double(M) * double(N) * double(K);
-
-    ptrdiff_t TargetThreadCount;
-
-    if (Complexity < double(MLAS_QGEMM_THREAD_COMPLEXITY * MlasPlatform.MaximumThreadCount)) {
-        TargetThreadCount = ptrdiff_t(Complexity / double(MLAS_QGEMM_THREAD_COMPLEXITY)) + 1;
-    } else {
-        TargetThreadCount = MlasPlatform.MaximumThreadCount;
-    }
-
-    ptrdiff_t MaximumThreadCount = MlasGetMaximumThreadCount(ThreadPool);
-
-    if (TargetThreadCount >= MaximumThreadCount) {
-        TargetThreadCount = MaximumThreadCount;
-    }
-
-    //
-    // Segment the operation across multiple threads.
-    //
-    // N.B. Currently, the operation is segmented as a 1D partition, which
-    // works okay for operations involving skinny matrices.
-    //
-
-    MLAS_GEMM_U8X8_WORK_BLOCK WorkBlock;
-
-    WorkBlock.Parameters = Parameters;
-
-    if (N > M) {
-
-        const size_t BlockedN = (N + MLAS_QGEMM_STRIDEN_THREAD_ALIGN - 1) /
-            MLAS_QGEMM_STRIDEN_THREAD_ALIGN;
-
-        if (size_t(TargetThreadCount) > BlockedN) {
-            TargetThreadCount = ptrdiff_t(BlockedN);
-        }
-
-        WorkBlock.ThreadCountM = 1;
-        WorkBlock.ThreadCountN = TargetThreadCount;
-
-    } else {
+  MlasGemmBatch(Shape, &DataParams, 1, ThreadPool);
+}
 
-        if (size_t(TargetThreadCount) > M) {
-            TargetThreadCount = ptrdiff_t(M);
+void
+MLASCALL
+MlasGemmBatch(const MLAS_GEMM_U8X8_SHAPE_PARAMS& Shape,
+              const MLAS_GEMM_U8X8_DATA_PARAMS* DataParams,
+              const size_t BatchN,
+              MLAS_THREADPOOL* ThreadPool) {
+  const size_t M = Shape.M;
+  const size_t N = Shape.N;
+  const size_t K = Shape.K;
+
+  // Segment the work for parallelization. This is a two dimentional work partition.
+  // The idea is to generate a group of not-too-small work chunks and let the thread
+  // pool load balance them
+
+  std::ptrdiff_t BlksPerGemm;
+  double cost;
+  MlasGemmSegWork(M, N, K, Shape.BIsSigned, BlksPerGemm, cost);
+
+  MLAS_THREADPOOL::TryParallelFor(
```

Comment:
need to #ifdef this for "defined(MLAS_NO_ONNXRUNTIME_THREADPOOL)".

Reply:
do we have existing setup for testing this? i.e. a way to compile MLAS outside of ORT?